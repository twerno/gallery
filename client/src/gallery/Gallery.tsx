import { IImageQueryRespBody } from '@shared/';

export interface IGalleryProps {
    images: IImageQueryRespBody;
}

export const Gallery = (props: IGalleryProps) => {

}
