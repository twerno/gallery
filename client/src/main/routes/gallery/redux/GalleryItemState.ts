import { IGiphyImage, IPixabayImage } from '@shared/';
import { IGalleryUrlQuery } from 'main/routes/Path';

export interface IPreviewGiphyImg extends IGiphyImage {
    imgProvider: 'giphy';
}

export interface IPreviewPixabyImg extends IPixabayImage {
    imgProvider: 'pixabay';
}

export type IPreviewImg = IPreviewGiphyImg | IPreviewPixabyImg;

export interface IGalleryStateMeta {
    pageIdx: number;
    limit?: {
        pixabayPages: number;
        giphyPages: number;
    }
    query?: IGalleryUrlQuery;
}

export interface IGalleryState {
    images: (IPreviewGiphyImg | IPreviewPixabyImg)[];
    loadingMeta: IGalleryStateMeta;

    errors?: string[];

    previewIdx?: number;
}

export interface IItemsLoadedActionPayload {
    images: (IPreviewGiphyImg | IPreviewPixabyImg)[];
    limit: {
        pixabayPages: number;
        giphyPages: number;
    }
}
