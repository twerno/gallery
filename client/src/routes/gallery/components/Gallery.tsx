import { ILocalGiphyGetImageReturnModel, ILocalPixabayGetImageReturnModel } from '@shared/';
import * as React from 'react';

import styles from '../gallery.module.css';
import { GiphyRenderer } from './GiphyRenderer';
import { LazyLoadMore } from './LazyLoadMore';
import { PixabyRenderer } from './PixabyRenderer';

export interface IGalleryProps {
    hasMorePages: boolean;
    canLoadMorePages: boolean;
    loadNextPageCallback: () => void;
    pages: Array<ILocalGiphyGetImageReturnModel | ILocalPixabayGetImageReturnModel>[] | undefined;
}

export const Gallery = (props: IGalleryProps) => {

    const placeholder = <Placeholder />;

    const images = props.pages?.map(page =>
        (page || []).map((images, pageIdx) => {
            switch (images.imgProvider) {
                case 'giphy':
                    return images.data.map(img => <GiphyRenderer placeholder={placeholder} image={img} key={img.id} />);

                case 'pixabay':
                    return images.hits.map(img => <PixabyRenderer placeholder={placeholder} image={img} key={img.id} />);

                default:
                    return <div key={pageIdx}>{`Unknown provider: "${(images as any).imgProvider}"`}</div>;
            }
        })
    );

    return (
        <div className={styles.wrapper}>
            <div className={styles.gallery}>
                {images}
                {props.hasMorePages && props.canLoadMorePages &&
                    <LazyLoadMore
                        placeholder={placeholder}
                        loadMoreCallback={props.loadNextPageCallback}
                        key={`load_next_page_${props.pages?.length}`}
                    />
                }
            </div>
        </div>
    );
}

const Placeholder = () => <div className={styles.galleryItemPlaceholder} />;
