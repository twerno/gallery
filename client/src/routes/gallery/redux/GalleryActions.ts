import { IGalleryUrlQuery } from 'routes/Path';

import { IPreviewGiphyImg, IPreviewPixabyImg } from './GalleryState';

export const GalleryQueryChangeAction = 'GalleryQueryChangeAction'
export const GalleryImagesLoadedAction = 'GalleryImagesLoadedAction'
export const GalleryLoadMoreAction = 'GallertLoadMoreAction';
export const GalleryLoadingErrors = 'GalleryLoadingErrors';
export const GallerySetPreview = 'GallerySetPreview';

export type GALLERY_ACTION_TYPE =
    typeof GalleryQueryChangeAction
    | typeof GalleryImagesLoadedAction
    | typeof GalleryLoadMoreAction
    | typeof GallerySetPreview
    ;

export interface IGalleryChangeQueryAction {
    type: typeof GalleryQueryChangeAction;
    data: {
        query: IGalleryUrlQuery;
    }
}


export interface IGalleryImagesLoadedAction {
    type: typeof GalleryImagesLoadedAction;
    data: {
        images: (IPreviewGiphyImg | IPreviewPixabyImg)[];
        limit: {
            pixabayPages: number;
            giphyPages: number;
        }
    }
}

export interface IGallertLoadMoreAction {
    type: typeof GalleryLoadMoreAction;
}

export interface IGallerySetErrorAction {
    type: typeof GalleryLoadingErrors;
    data: {
        errors: string[];
    }
}

export interface IGallerySetPreviewAction {
    type: typeof GallerySetPreview;
    data: {
        previewIdx?: number;
    };
}

export type GalleryActions = IGalleryChangeQueryAction
    | IGalleryImagesLoadedAction
    | IGallertLoadMoreAction
    | IGallerySetErrorAction
    | IGallerySetPreviewAction
    ;
