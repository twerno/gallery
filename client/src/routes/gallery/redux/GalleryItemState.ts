import { IGiphyImage, IPixabayImage } from '@shared/';
import { IGalleryUrlQuery } from 'routes/Path';

export interface IPreviewGiphyImg extends IGiphyImage {
    imgProvider: 'giphy';
}

export interface IPreviewPixabyImg extends IPixabayImage {
    imgProvider: 'pixabay';
}

export interface IGalleryStateMeta {
    pageIdx: number;
    limit?: {
        pixabayPages: number;
        giphyPages: number;
    }
    query?: IGalleryUrlQuery;
}

export interface IGalleryState {
    images: (IPreviewGiphyImg | IPreviewPixabyImg)[];
    loadingMeta: IGalleryStateMeta;

    errors?: string[];

    previewIdx?: number;
}

export interface IItemsLoadedActionPayload {
    images: (IPreviewGiphyImg | IPreviewPixabyImg)[];
    limit: {
        pixabayPages: number;
        giphyPages: number;
    }
}
