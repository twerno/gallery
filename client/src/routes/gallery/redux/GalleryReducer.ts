import {
    GalleryActions,
    GalleryImagesLoadedAction,
    GalleryLoadingErrors,
    GalleryLoadMoreAction,
    GalleryQueryChangeAction,
    GallerySetPreview,
} from './GalleryActions';
import { IGalleryState } from './GalleryState';

const initialState: IGalleryState = {
    images: [],
    loadingMeta: {
        pageIdx: 0
    }
};

export function galleryReducer(
    state = initialState,
    action: GalleryActions
): IGalleryState {
    switch (action.type) {

        case GalleryQueryChangeAction:
            return {
                images: [],
                loadingMeta: {
                    query: { ...action.data.query },
                    pageIdx: 0,
                    limit: undefined
                }
            }

        case GalleryImagesLoadedAction:
            return {
                ...state,
                images: [...state.images, ...action.data.images],
                loadingMeta: {
                    ...state.loadingMeta,
                    limit: action.data.limit,
                },
                errors: undefined
            }

        case GalleryLoadMoreAction:
            return {
                ...state,
                loadingMeta: {
                    ...state.loadingMeta,
                    pageIdx: state.loadingMeta.pageIdx + 1
                },
                errors: undefined
            }

        case GalleryLoadingErrors:
            return {
                ...state,
                errors: [...action.data.errors]
            }

        case GallerySetPreview:
            return {
                ...state,
                previewIdx: action.data.previewIdx
            }

        default:
            return state
    }
}