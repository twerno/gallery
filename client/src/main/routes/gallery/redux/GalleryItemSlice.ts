import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { IStringMap } from '@shared/';
import { IGalleryUrlQuery } from 'main/routes/Path';

import { IGalleryState, IItemsLoadedActionPayload } from './GalleryItemState';

const initialState: IGalleryState = {
    images: [],
    loadingMeta: {
        pageIdx: 0
    }
}

export const galleryItemSlice = createSlice({
    name: 'gallery',
    initialState,
    reducers: {

        newQuery(state, action: PayloadAction<IGalleryUrlQuery>): IGalleryState {
            return {
                images: [],
                loadingMeta: {
                    pageIdx: 0,
                    query: { ...action.payload }
                }
            };
        },

        itemsLoaded(state: IGalleryState, action: PayloadAction<IItemsLoadedActionPayload>): IGalleryState {
            const imageIds: IStringMap<any> = {};
            state.images.forEach(img => imageIds[img.id] = null);
            const newImages = action.payload.images.filter(img => imageIds[img.id] === undefined);
            return {
                ...state,
                images: [...state.images, ...newImages],
                loadingMeta: {
                    ...state.loadingMeta,
                    limit: { ...action.payload.limit }
                }
            };
        },

        loadNextPage(state): IGalleryState {
            return {
                ...state,
                loadingMeta: {
                    ...state.loadingMeta,
                    pageIdx: state.loadingMeta.pageIdx + 1
                }
            };
        },

        loadingErrors(state, action: PayloadAction<string[]>): IGalleryState {
            return {
                ...state,
                errors: [...action.payload]
            };
        },

        setPreview(state, action: PayloadAction<number | undefined>): IGalleryState {
            return {
                ...state,
                previewIdx: action.payload
            };
        },

        previewMove(state, action: PayloadAction<'prev' | 'next'>): IGalleryState {

            if (state.previewIdx === undefined) {
                return state;
            }

            const nextPreviewIdx = action.payload === 'prev' && state.previewIdx > 0
                ? state.previewIdx - 1
                : action.payload === 'next' && state.previewIdx < state.images.length
                    ? state.previewIdx + 1
                    : state.previewIdx;

            return {
                ...state,
                previewIdx: nextPreviewIdx
            };
        },

    }
});
