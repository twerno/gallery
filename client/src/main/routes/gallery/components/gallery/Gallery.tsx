import * as React from 'react';
import { IPreviewGiphyImg, IPreviewPixabyImg } from '../../redux/GalleryItemState';
import { GalleryItemGiphy } from './GalleryItemGiphy';
import { GalleryItemPixaby } from './GalleryItemPixaby';
import { GalleryContainer } from './GalleryStyles';

export interface IGalleryProps {
    disabled: boolean;
    images: (IPreviewGiphyImg | IPreviewPixabyImg)[];
}

export const Gallery = (props: IGalleryProps) => {

    const items = props.images?.map(
        (img, idx) =>
            img.imgProvider === 'giphy'
                ? <GalleryItemGiphy
                    image={img}
                    imageIdx={idx}
                    key={`giphy_${img.id}`}
                    disabled={props.disabled}
                />
                : <GalleryItemPixaby
                    image={img}
                    imageIdx={idx}
                    key={`pixabay_${img.id}`}
                    disabled={props.disabled}
                />
    );

    return (
        <GalleryContainer>
            {items}
        </GalleryContainer>
    );
};
