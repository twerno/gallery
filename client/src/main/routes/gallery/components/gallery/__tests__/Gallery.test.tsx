import '@testing-library/jest-dom/extend-expect';
import 'jest-styled-components';
import { IPreviewGiphyImg, IPreviewPixabyImg } from 'main/routes/gallery/redux/GalleryItemState';
import * as React from 'react';
import { componentBasicTests } from 'test/componentBasicTests';
import { Gallery } from '../Gallery';

describe('Gallery', () => {

    componentBasicTests(
        "empty Gallery - disabled: false",
        <Gallery images={[]} disabled={false} />
    );

    componentBasicTests(
        "empty Gallery - disabled: true",
        <Gallery images={[]} disabled={true} />
    );

    const images: (IPreviewGiphyImg | IPreviewPixabyImg)[] = [
        {
            imgProvider: 'giphy',
            embed_url: 'embed_url',
            id: 'id',
            images: {
                fixed_height_small_still: {
                    height: '200',
                    url: 'http://example.com.not/image.jpg',
                    width: '200'
                },
                original: {
                    height: '300',
                    url: 'http://example.com.not/image.jpg',
                    width: '300'
                }
            },
            title: 'title',
            type: 'type'
        }
    ];

    componentBasicTests(
        "Gallery - disabled: false",
        <Gallery images={images} disabled={false} />
    );

    componentBasicTests(
        "Gallery - disabled: true",
        <Gallery images={images} disabled={true} />
    );

});

