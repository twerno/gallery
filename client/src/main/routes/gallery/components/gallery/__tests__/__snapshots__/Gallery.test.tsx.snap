// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gallery [Gallery - disabled: false] Component basic tests matches the snapshot 1`] = `
<styled.div>
  <Component
    disabled={false}
    imageIdx={0}
    key="giphy_id"
    responsivePicture={
      Object {
        "alt": "title",
        "defaultImgSrc": "http://example.com.not/image.jpg",
        "responsiveSrcList": Array [
          Object {
            "media": "(min-width: 600px)",
            "src": undefined,
          },
        ],
      }
    }
  />
</styled.div>
`;

exports[`Gallery [Gallery - disabled: true] Component basic tests matches the snapshot 1`] = `
<styled.div>
  <Component
    disabled={true}
    imageIdx={0}
    key="giphy_id"
    responsivePicture={
      Object {
        "alt": "title",
        "defaultImgSrc": "http://example.com.not/image.jpg",
        "responsiveSrcList": Array [
          Object {
            "media": "(min-width: 600px)",
            "src": undefined,
          },
        ],
      }
    }
  />
</styled.div>
`;

exports[`Gallery [empty Gallery - disabled: false] Component basic tests matches the snapshot 1`] = `<styled.div />`;

exports[`Gallery [empty Gallery - disabled: true] Component basic tests matches the snapshot 1`] = `<styled.div />`;
